src/java/org/apache/cassandra/config/DataStorageSpec.java

```diff
@@ -38,221 +36,422 @@
+         * Creates a {@code DataStorageSpec.LongMebibytesBound} of the specified amount in the specified unit.
+         *
+         * @param quantity where quantity shouldn't be bigger than Long.MAX_VALUE - 1 in mebibytes
+         * @param unit in which the provided quantity is
+         */
+        public LongMebibytesBound(long quantity, DataStorageUnit unit)
+        {
+            super(quantity, unit, MEBIBYTES, Long.MAX_VALUE, quantity + unit.symbol);
+        }
+
+        /**
+         * Creates a {@code DataStorageSpec.LongMebibytesBound} of the specified amount in mebibytes.
+         *
+         * @param mebibytes where mebibytes shouldn't be bigger than Long.MAX_VALUE-1
+         */
+        public LongMebibytesBound(long mebibytes)
+        {
+            this(mebibytes, MEBIBYTES);
+        }
+
+        /**
+         * @return the amount of data storage in bytes
+         */
+        public long toBytes()
+        {
+            return unit().toBytes(quantity());
+        }
+
+        /**
+         * @return the amount of data storage in kibibytes
+         */
+        public long toKibibytes()
+        {
+            return unit().toKibibytes(quantity());
+        }
+
+        /**
+         * @return the amount of data storage in mebibytes
+         */
+        public long toMebibytes()
+        {
+            return unit().toMebibytes(quantity());
+        }
     }
 
     /**
-     * Returns the amount of data storage in gibibytes as an {@code int}
-     *
-     * @return the amount of data storage in gibibytes or {@code Integer.MAX_VALUE} if the number of gibibytes is too large.
+     * Represents a data storage quantity used for Cassandra configuration. The bound is [0, Integer.MAX_VALUE) in mebibytes.
+     * If the user sets a different unit - we still validate that converted to mebibytes the quantity will not exceed
+     * that upper bound. (CASSANDRA-17571)
      */
-    public int toGibibytesAsInt()
+    public final static class IntMebibytesBound extends DataStorageSpec
     {
-        return Ints.saturatedCast(toGibibytes());
-    }
-
-    @Override
-    public int hashCode()
-    {
-        return Objects.hash(unit.toKibibytes(quantity));
-    }
+        /**
+         * Creates a {@code DataStorageSpec.IntMebibytesBound} of the specified amount.
+         *
+         * @param value the data storage
+         */
+        public IntMebibytesBound(String value)
+        {
+            super(value, MEBIBYTES, Integer.MAX_VALUE);
+        }
 
-    @Override
-    public boolean equals(Object obj)
-    {
-        if (this == obj)
-            return true;
+        /**
+         * Creates a {@code DataStorageSpec.IntMebibytesBound} of the specified amount in the specified unit.
+         *
+         * BE CAREFUL, IF YOU DECIDE TO USE UNIT BYTES OR KIBIBYTES, SET A NUMBER THAT WILL NOT LEAD TO LOSS OF PRECISION DURING CONVERSION
+         * TO MEBIBYTES. WE GUARD FOR THIS IN THE PREVIOUS CONSTRUCTOR BUT NOT THIS ONE
```

Comment:
Nit: we can reformat to avoid exceeding the line width:
```suggestion
         * BE CAREFUL, IF YOU DECIDE TO USE UNIT BYTES OR KIBIBYTES, SET A NUMBER THAT WILL NOT LEAD TO LOSS OF
         * PRECISION DURING CONVERSION TO MEBIBYTES. WE GUARD FOR THIS IN THE PREVIOUS CONSTRUCTOR BUT NOT THIS ONE.
```

Reply:
Actually we validate now this constructor, so I had to remove this too, same as others. I feel I messed up commits as I remember walking that one and removing it everywhere by searching for it in all files.... I will double check what happened